// @cloudscape-design/theming-runtime/shared/declaration/abstract.js
import { getThemeModeByState, isOptionalState } from '../theme/utils';
import { entries } from '../utils';
/**
 * Common used helpers by stylesheet creator
 */
var AbstractCreator = /** @class */ (function () {
    function AbstractCreator(ruleCreator, propertiesMap) {
        this.ruleCreator = ruleCreator;
        this.propertiesMap = propertiesMap;
    }
    AbstractCreator.forEachOptionalModeState = function (theme, func) {
        Object.keys(theme.modes).forEach(function (key) {
            var mode = theme.modes[key];
            entries(mode.states).forEach(function (_a) {
                var stateKey = _a[0], state = _a[1];
                if (isOptionalState(state)) {
                    func(mode, stateKey);
                }
            });
        });
    };
    AbstractCreator.forEachContext = function (theme, func) {
        Object.keys(theme.contexts).forEach(function (key) {
            var context = theme.contexts[key];
            func(context);
        });
    };
    AbstractCreator.forEachContextWithinOptionalModeState = function (theme, func) {
        AbstractCreator.forEachOptionalModeState(theme, function (mode, stateKey) {
            AbstractCreator.forEachContext(theme, function (context) {
                func(context, mode, stateKey);
            });
        });
    };
    AbstractCreator.appendRuleToStylesheet = function (stylesheet, rule, path) {
        if (rule.size()) {
            stylesheet.appendRuleWithPath(rule, path);
        }
    };
    /**
     * Finds stylesheet rule by selector or selector config.
     */
    AbstractCreator.prototype.findRule = function (stylesheet, config) {
        return stylesheet.findRule(typeof config === 'string' ? config : this.ruleCreator.selectorFor(config));
    };
    /**
     * Finds stylesheet rule or throws.
     *
     * Rules are creating in reverse order to dependency, which is why we should always have a rule.
     */
    AbstractCreator.prototype.findExpectedRule = function (stylesheet, config) {
        var rule = this.findRule(stylesheet, config);
        if (!rule) {
            throw new Error("No rule for selector ".concat(JSON.stringify(config), " found"));
        }
        return rule;
    };
    /**
     * Resolves the optional mode state a context inherits from (e.g. the "dark" mode) to deduplicate output.
     */
    AbstractCreator.prototype.findInheritedModeState = function (theme, _a) {
        var inheritsMode = _a.inheritsMode;
        if (!inheritsMode) {
            return null;
        }
        var mode = getThemeModeByState(theme, inheritsMode);
        if (!mode) {
            return null;
        }
        var state = mode.states[inheritsMode];
        if (!isOptionalState(state)) {
            return null;
        }
        var media = state.media;
        var selector = this.ruleCreator.selectorFor({ global: [theme.selector, state.selector] });
        return { state: inheritsMode, mode: mode, media: media, selector: selector };
    };
    /**
     * Resolves the optional mode state of the parent theme context.
     */
    AbstractCreator.prototype.findParentInheritedModeState = function (theme, contextId) {
        var parentContext = theme.contexts[contextId];
        return parentContext ? this.findInheritedModeState(theme, parentContext) : null;
    };
    /**
     * Appends context selector(s) to the inherited mode selector (as comma aliases).
     */
    AbstractCreator.prototype.registerInheritedContextAliases = function (stylesheet, inheritedModeRule, configs) {
        var _this = this;
        var aliasSelectors = Array.from(new Set(configs.map(function (config) { return _this.ruleCreator.selectorFor(config); })));
        aliasSelectors.forEach(function (aliasSelector) { return stylesheet.registerInheritedAlias(aliasSelector, inheritedModeRule); });
    };
    return AbstractCreator;
}());
export { AbstractCreator };
