// @cloudscape-design/theming-runtime/shared/declaration/stylesheet.js
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
/**
 * Allows building a stylesheet by appending selectors and declarations. A call of
 * `toString` will return CSS.
 */
var Stylesheet = /** @class */ (function () {
    function Stylesheet() {
        this.rulesMap = new Map();
        this.paths = new Map();
        this.counter = 0;
        this.inheritedAliases = [];
    }
    Stylesheet.prototype.appendRule = function (rule) {
        this.rulesMap.set(rule.selector, [rule, this.counter++]);
    };
    Stylesheet.prototype.appendRuleWithPath = function (rule, path) {
        // When a global theme (body/:root/html) has a context, the descendant form
        // { global: [body], local: [.ctx] } and the same-element form
        // { global: [body, .ctx] } both resolve to the same selector string because
        // the global selector is stripped. Silently skip the duplicate so that
        // rulesMap and paths stay in sync (one entry each per selector).
        if (this.rulesMap.has(rule.selector)) {
            return;
        }
        this.rulesMap.set(rule.selector, [rule, this.counter++]);
        this.paths.set(rule, path);
    };
    Stylesheet.prototype.removeRule = function (rule) {
        this.rulesMap["delete"](rule.selector);
        this.paths["delete"](rule);
    };
    Stylesheet.prototype.findRule = function (selector) {
        var ruleOrUndefined = this.rulesMap.get(selector);
        return ruleOrUndefined === null || ruleOrUndefined === void 0 ? void 0 : ruleOrUndefined[0];
    };
    Stylesheet.prototype.getPath = function (rule) {
        var path = this.paths.get(rule);
        if (!path) {
            throw new Error("No path for rule with selector: ".concat(rule.selector));
        }
        return path;
    };
    Stylesheet.prototype.getAllRules = function () {
        var rules = [];
        this.paths.forEach(function (_, key) { return rules.push(key); });
        return rules;
    };
    /**
     * Records that an inheriting context's selector should be appended (as a comma alias)
     * onto an inherited mode state's rule. Applied as a final step by the transformer,
     * after all rule lookups are done, so it never interferes with `findRule`.
     */
    Stylesheet.prototype.registerInheritedAlias = function (aliasSelector, rule) {
        this.inheritedAliases.push({ aliasSelector: aliasSelector, rule: rule });
    };
    /**
     * Appends each inheriting context's selector as a comma alias onto its inherited
     * mode state's rule. This makes the inherited mode values apply to the context
     * (e.g. `.dark-mode, .top-navigation { ... }`) without duplicating them, while
     * the context's standalone rule keeps only the tokens that differ from the mode.
     */
    Stylesheet.prototype.applyInheritedAliases = function () {
        var present = new Set(this.getAllRules());
        this.inheritedAliases.forEach(function (_a) {
            var aliasSelector = _a.aliasSelector, rule = _a.rule;
            if (!present.has(rule)) {
                return;
            }
            rule.selector = "".concat(rule.selector, ",").concat(aliasSelector);
        });
    };
    /**
     * Merges adjacent rules with identical declarations into a single comma-separated selector.
     *
     * For example:
     * .a { --color-background: red; }
     * .b { --color-background: red; }
     *
     * becomes:
     * .a,
     * .b { --color-background: red; }
     */
    Stylesheet.prototype.mergeSelectors = function () {
        var rules = this.getAllRules();
        var i = 1;
        while (i < rules.length) {
            var prev = rules[i - 1];
            var curr = rules[i];
            if (prev.media === curr.media && prev.printAllDeclarations() === curr.printAllDeclarations()) {
                prev.selector = "".concat(prev.selector, ",").concat(curr.selector);
                this.removeRule(curr);
                rules.splice(i, 1);
            }
            else {
                i++;
            }
        }
    };
    /**
     * @returns CSS
     */
    Stylesheet.prototype.toString = function (layer) {
        var result = asValuesArray(this.rulesMap).map(function (rule) { return rule.toString(); });
        return layer ? "@layer ".concat(layer, " {\n").concat(result.join('\n'), "\n}") : result.join('\n');
    };
    return Stylesheet;
}());
export default Stylesheet;
var Rule = /** @class */ (function () {
    function Rule(selector, media, isContext) {
        if (isContext === void 0) { isContext = false; }
        this.declarationsMap = new Map();
        this.counter = 0;
        this.selector = selector;
        this.media = media;
        this.isContext = isContext;
    }
    Rule.prototype.appendDeclaration = function (declaration) {
        this.declarationsMap.set(declaration.property, [declaration, this.counter++]);
    };
    Rule.prototype.clear = function () {
        this.declarationsMap = new Map();
        this.counter = 0;
    };
    Rule.prototype.getAllDeclarations = function () {
        return asValuesArray(this.declarationsMap);
    };
    Rule.prototype.printAllDeclarations = function () {
        return this.getAllDeclarations()
            .map(function (decl) { return decl.toString(); })
            .join('\n\t');
    };
    Rule.prototype.size = function () {
        return this.declarationsMap.size;
    };
    Rule.prototype.toString = function () {
        var rule = "".concat(this.selector, "{\n\t").concat(this.printAllDeclarations(), "\n}");
        if (this.media) {
            return "@media ".concat(this.media, " {").concat(rule, "}");
        }
        return rule;
    };
    Rule.prototype.isModeRule = function () {
        return !!this.media;
    };
    Rule.prototype.isContextRule = function () {
        return this.isContext;
    };
    return Rule;
}());
export { Rule };
var Declaration = /** @class */ (function () {
    function Declaration(property, value) {
        this.property = property;
        this.value = value;
    }
    Declaration.prototype.toString = function () {
        return "".concat(this.property, ":").concat(this.value, ";");
    };
    return Declaration;
}());
export { Declaration };
function asValuesArray(map) {
    var tmp = [];
    map.forEach(function (_a) {
        var item = _a[0], position = _a[1];
        return tmp.push([item, position]);
    });
    tmp.sort(function (_a, _b) {
        var itemA = _a[0], posA = _a[1];
        var itemB = _b[0], posB = _b[1];
        return posA - posB;
    });
    return tmp.map(function (_a) {
        var item = _a[0];
        return item;
    });
}
