// @cloudscape-design/theming-runtime/shared/declaration/transformer.js
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import { entries } from '../utils';
import { Declaration } from './stylesheet';
import { getFirstSelector, isGlobalSelector } from '../styles/selector';
import { getReferencedVar } from './utils';
var MinimalTransformer = /** @class */ (function () {
    function MinimalTransformer() {
    }
    MinimalTransformer.prototype.transform = function (stylesheet) {
        var rules = stylesheet.getAllRules();
        var rulesWithPath = rules.map(function (rule) { return ({
            rule: rule,
            path: stylesheet.getPath(rule)
        }); });
        var sorted = rulesWithPath.sort(function (_a, _b) {
            var pathA = _a.path;
            var pathB = _b.path;
            return pathA.length - pathB.length;
        });
        sorted.forEach(function (_a) {
            var rule = _a.rule, path = _a.path;
            if (path.length === 0) {
                // Root rule nothing to see here.
                return;
            }
            var resolvedParent = {};
            for (var i = path.length - 1; i >= 0; i--) {
                var parent_1 = path[i];
                var declarations = parent_1.getAllDeclarations();
                declarations.forEach(function (decl) {
                    resolvedParent[decl.property] = decl.value;
                });
            }
            var ruleValue = rule.getAllDeclarations().reduce(function (acc, decl) {
                acc[decl.property] = decl.value;
                return acc;
            }, {});
            var diff = difference(resolvedParent, ruleValue);
            // CSS variables with nested var() references need special handling for non-global selectors.
            // Even if the selector doesn't explicitly show a descendant combinator (like `.navigation`),
            // it will be a descendant of `body` in the DOM. When a descendant overrides a token,
            // tokens that reference it must be re-output, otherwise they resolve in the parent context.
            //
            // However, for plain mode rules (media query, no context selector) we skip tokens that have
            // identical values to their parent, even if referenced variables are overridden: these inherit
            // properly via the natural CSS variable cascade. Context rules (including inheriting context
            // rules that also carry a media query) are always re-output, hence the explicit isContextRule.
            var firstSelector = getFirstSelector(rule.selector);
            var isModeRule = rule.isModeRule();
            var isContextRule = rule.isContextRule();
            if (isGlobalSelector(firstSelector)) {
                rule.clear();
                entries(diff).forEach(function (_a) {
                    var property = _a[0], value = _a[1];
                    return rule.appendDeclaration(new Declaration(property, value));
                });
                if (rule.size() === 0) {
                    stylesheet.removeRule(rule);
                }
                return;
            }
            var isOverridden = function (varName, visited) {
                if (visited === void 0) { visited = new Set(); }
                if (visited.has(varName))
                    return false;
                visited.add(varName);
                var isDirectlyOverridden = varName in ruleValue && varName in resolvedParent && ruleValue[varName] !== resolvedParent[varName];
                if (isDirectlyOverridden)
                    return true;
                var referencedVar = varName in ruleValue ? getReferencedVar(ruleValue[varName]) : null;
                return referencedVar ? isOverridden(referencedVar, visited) : false;
            };
            Object.keys(ruleValue).forEach(function (property) {
                var referencedVar = getReferencedVar(ruleValue[property]);
                if (!referencedVar || !isOverridden(referencedVar))
                    return;
                // This decides whether a token that references an overridden var
                // can be left out and resolved via the natural CSS variable cascade.
                // - Plain mode rules: yes - identical values inherit fine, so skip re-emitting.
                // - Context rules: no - a context rule must resolve in the context's own DOM scope.
                var canInherit = isModeRule && !isContextRule && ruleValue[property] === resolvedParent[property];
                if (canInherit)
                    return;
                if (!(property in diff)) {
                    diff[property] = ruleValue[property];
                }
            });
            rule.clear();
            entries(diff).forEach(function (_a) {
                var property = _a[0], value = _a[1];
                return rule.appendDeclaration(new Declaration(property, value));
            });
            if (rule.size() === 0) {
                stylesheet.removeRule(rule);
            }
        });
        // Inherited aliases are applied as near-last step so that findRule lookups work correctly.
        stylesheet.applyInheritedAliases();
        // Selectors are merged after aliases resolution so that final selectors are compared.
        stylesheet.mergeSelectors();
        return stylesheet;
    };
    return MinimalTransformer;
}());
export { MinimalTransformer };
function difference(mapA, mapB) {
    var diff = {};
    Object.keys(mapA).forEach(function (key) {
        if (mapA[key] !== mapB[key] && mapB[key] !== undefined) {
            diff[key] = mapB[key];
        }
    });
    Object.keys(mapB).forEach(function (key) {
        if (mapA[key] !== mapB[key] && mapB[key] !== undefined) {
            diff[key] = mapB[key];
        }
    });
    return diff;
}
